common/searchattribute/manager.go

```diff
@@ -112,28 +113,37 @@ func (m *managerImpl) GetSearchAttributes(
 }
 
 func (m *managerImpl) needRefreshCache(saCache cache, forceRefreshCache bool, now time.Time) bool {
-	return forceRefreshCache || saCache.lastRefresh.Add(cacheRefreshInterval).Before(now)
+	return forceRefreshCache || saCache.expireOn.Before(now)
 }
 
 func (m *managerImpl) refreshCache(saCache cache, now time.Time) (cache, error) {
 	clusterMetadata, err := m.clusterMetadataManager.GetClusterMetadata()
 	if err != nil {
-		if _, isNotFoundErr := err.(*serviceerror.NotFound); !isNotFoundErr {
+		switch err.(type) {
+		case *serviceerror.NotFound:
+			// NotFound means cluster metadata was never persisted and custom search attributes are not defined.
+			// Ignore the error.
+			saCache.expireOn = now.Add(cacheRefreshInterval)
+		case *serviceerror.Unavailable:
+			// If persistence is Unavailable, ignore the error and use existing cache for cacheRefreshIfUnavailableInterval.
+			saCache.expireOn = now.Add(cacheRefreshIfUnavailableInterval)
+		default:
 			return saCache, err
 		}
+		m.cache.Store(saCache)
```

Comment:
why do you need to do this here? and is saCache ever possible to be nil?

Reply:
1. Because value of `saCache.expireOn` has changed.
2. It can't be `nil` because it is `struct` passed by value.